### PR TITLE
Support for `@validateOn`/`@revalidateOn` options

### PR DESCRIPTION
**Pull request by simonihmig**
As an iteration over #21, this adds the options to validate and revalidate (when initial validation had errors) not only on submit, but also on blur/focusout and change.

It supersedes #24, and still relies on option 1 discussed in that draft PR. In a future iteration it is expected that we had further primitives for custom controls, like "option 4" of yielding a modifier to capture focusout/change event, or yielding more validation state like `field.isValid`. But this can all be done on top of this PR.

### CODE REVIEW

File: ember-headless-form/src/components/headless-form.ts [hunk 1]
```diff
@@ -146,4 +219,36 @@ export default class HeadlessFormComponent<
...
+
+    if (name) {
+      const field = this.fields.get(name as FormKey<FormData<DATA>>);
+
+      if (field) {
+        this.lastValidationResult = await this.validate();
+        field.validationEnabled = true;
+      }
+    } else {
+      // @todo how to handle custom controls that don't emit focusout/change events from native form controls?
```
Comment (simonihmig) [inline ember-headless-form/src/components/headless-form.ts:236]: The DX should be improved here by emitting a warning, by a future PR (see PR description) that provides additional primitives for better custom controls support, and mentions these primitives in the warning. See keeping this as-is for now...

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 2]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+        );
+      });
+
+      test('validation errors are exposed as field.errors on blur', async function (assert) {
+        const data: TestFormData = { firstName: 'Tony', lastName: 'Foo' };
+
+        await render(<template>
+          <HeadlessForm
+            @data={{data}}
+            @validateOn="blur"
```
Comment (nicolechung) [inline test-app/tests/integration/components/headless-form-validation-test.gts:949]: Question: in glint does this auto hint `blur` or `change`?

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 3]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+            </form.field>
+            <button type="submit" data-test-submit>Submit</button>
+          </HeadlessForm>
+        </template>);
+
+        await fillIn('[data-test-first-name]', 'Foo');
+
+        assert.false(
+          validateCallback.called,
+          '@validate is not called while typing'
```
Comment (nicolechung) [inline test-app/tests/integration/components/headless-form-validation-test.gts:1167]: ✅  nice!

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 4]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+        assert.true(
+          validateCallback.getCall(1).calledWith('Tony', 'firstName', {
+            ...data,
+            firstName: 'Tony',
+          }),
+          '@validate is called with form data'
+        );
+      });
+
+      test('validation errors are exposed as field.errors on blur', async function (assert) {
```
Comment (nicolechung) [inline test-app/tests/integration/components/headless-form-validation-test.gts:1281]: Suggestion: `'validation errors are exposed as field.errors on blur'` -> `'validation errors are exposed as field.errors on @revalidateOn="blur'`

**Review by nicolechung — COMMENTED** [review]
So far looking GREAT 💯 ...I have a bunch of questions / suggestions...but very excited to see how easy this is to use

File: ember-headless-form/src/components/headless-form.ts [hunk 5]
```diff
@@ -146,4 +219,36 @@ export default class HeadlessFormComponent<
...
+  }
+
+  @action
+  async handleFieldRevalidation(e: Event): Promise<void> {
+    const { target } = e;
+    const { name } = target as HTMLInputElement;
+
+    if (name) {
+      if (this.showErrorsFor(name as FormKey<FormData<DATA>>)) {
+        this.lastValidationResult = await this.validate();
```
Comment (NullVoxPopuli) [inline ember-headless-form/src/components/headless-form.ts:247]: interacting with `this` after an await isn't safe 🙈 

I keep meaning to work on [this lint](https://github.com/ember-cli/eslint-plugin-ember/pull/1421) but... priorities 🙃 

be sure to `if (isDestroyed(this) || isDestroying(this)) return` after an await

File: ember-headless-form/src/components/headless-form.ts [hunk 6]
```diff
@@ -67,6 +85,34 @@ export default class HeadlessFormComponent<
...
+    return validateOn === 'submit'
+      ? // no need for dynamic validation, as validation always happens on submit
+        undefined
+      : // our component API expects "blur", but the actual blur event does not bubble up, so we use focusout internally instead
+      validateOn === 'blur'
+      ? 'focusout'
+      : validateOn;
+  }
+
+  get fieldRevalidationEvent(): 'focusout' | 'change' | undefined {
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply ember-headless-form/src/components/headless-form.ts:125]: Added, and a bunch of more JSDoc annotations along that!

This was good to call out, thank you! Indeed there are different things playing together here, which were clear to me as I was working on the things directly for many hours, but which might not be obvious on a review or when working in the codebase and not being super familiar with it. So, yeah, I must remind myself to explain things better here!

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 7]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+        assert.false(
+          validateCallback.called,
+          '@validate is not called while typing'
+        );
+
+        await blur('[data-test-first-name]');
+
+        assert.true(
+          validateCallback.calledWith({ ...data, firstName: 'Foo' }),
+          '@validate is called with form data'
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply test-app/tests/integration/components/headless-form-validation-test.gts]: Applied most of your assertion suggestions here, and at all the other places were similar things were tested! Thanks!

File: test-app/tests/integration/components/headless-form-validation-test.gts (hunk 2 above)
Comment (simonihmig) [reply test-app/tests/integration/components/headless-form-validation-test.gts:949]: Yeah, so I just tested this, and Glint calls out an error if you pass anything but the expected values, so here `blur`, `change` or `submit`. But it actually did not provide something like autocomplete (like typing `b` and suggesting `blur`), which I was expecting. I am not sure why that is, or if I am expecting too much 🤔 

But at least the types seem correct from our end seem correct...

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 8]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+
+        assert.true(
+          validateCallback.calledTwice,
+          '@validate is called for revalidation on blur'
+        );
+        assert.true(
+          validateCallback
+            .getCall(1)
+            .calledWith({ ...data, firstName: 'Tony' }),
+          '@validate is called with form data'
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply test-app/tests/integration/components/headless-form-validation-test.gts]: Yeah, that is what is entered basically, which updates on every change. 

And no, it is _not_ the native [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) API, just a simple POJO. But that's deliberately the case, to support custom control use cases that we touched already, like the 3-select-datepicker, which should map to a single `date` field instead of 3 individual ones, which is what a native `new FormData(formElement)` would give you.

If there comes up a need to actually provide a `FormData` instance instead of that POJO, we can do that easily by creating it from that POJO. But I think we should keep the functionality to have form data that does not map 1:1 to native form controls.

File: test-app/tests/integration/components/headless-form-validation-test.gts (hunk 4 above)
Comment (simonihmig) [reply test-app/tests/integration/components/headless-form-validation-test.gts:1281]: I would disagree here, because
* we have that wrapped in a `module()` that already makes clear that we are testing the `@revalidateOn="blur'` scenario
* the `on blur` refers to what the user is doing here, not what the dev is passing to the component

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 9]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+        assert
+          .dom('[data-test-last-name-errors]')
+          .doesNotExist(
+            'validation errors are not rendered for untouched fields'
+          );
+      });
+    });
+  });
+
+  module(`@revalidateOn`, function () {
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply test-app/tests/integration/components/headless-form-validation-test.gts:1138]: I have mixed feelings about this, because it is hard to draw the boundaries here for what should go into what testing file, as all these things belong together and can be combined. For example I don't have test coverage here for all the combinations, like `@validateOn="blur" @revalidateOn="change"`. But this would be a good and valid test scenario, and into which test file should be put it then?

Also we will probably add additional validation-related features, like this `captureEvents` modifier (aka "option 4"), which we might want to test with a particular `@revalidateOn` config. So should this then be in the `validation` file, the `revalidate` file or its own? 

Idk, like I am not opposed to this, as I said I am just unsure how to draw the boundaries and not cause more confusion than we wanted to prevent by splitting this up...

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 10]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+        );
+        assert.true(
+          validateCallback
+            .getCall(1)
+            .calledWith({ ...data, firstName: 'Tony' }),
+          '@validate is called with form data'
+        );
+      });
+
+      test('field validation callback is called on blur', async function (assert) {
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply test-app/tests/integration/components/headless-form-validation-test.gts:1209]: see above

File: test-app/tests/integration/components/headless-form-validation-test.gts [hunk 11]
```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
...
+
+        await render(<template>
+          <HeadlessForm
+            @data={{data}}
+            @revalidateOn="change"
+            @validate={{validateFormCallbackSync}}
+            as |form|
+          >
+            <form.field @name="firstName" as |field|>
+              <field.label>First Name</field.label>
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply test-app/tests/integration/components/headless-form-validation-test.gts:1541]: Great suggestion! Yes, we should do that, I'll keep that in my mind (hopefully, scold me when I forget :wink:). But I would add that in a future PR that also adds the primitive to support that use case, which we currently don't have yet.

File: ember-headless-form/src/components/-private/types.ts [hunk 12]
```diff
@@ -55,7 +55,7 @@ export type FieldValidateCallback<
  * Internal structure to track used fields
  * @private
  */
-export interface FieldData<
+export interface FieldRegistrationData<
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply ember-headless-form/src/components/-private/types.ts:58]: I was splitting this up into two concerns, that's why I renamed this.

So there is information passed form the field component to the parent form component via the `registerField` action. Basically the field register itself with its parent form, so the form has data available that the dev passed to the field. Currently this is just the _field_ level validation callback, which the form needs to have available. This is what the `FieldRegistrationData` contains.

Then in this PR there is a new [`FieldData` class being added](https://github.com/CrowdStrike/ember-headless-form/pull/25/files#diff-71dca24e838c62279dc22478137e692716eddbccaf782f3901bfda818c5e67daR54-R71), which gets created when registering the field, so it uses `FieldRegistrationData`, but it also contains more than that (the `validationEnabled` tracked property), which the field does not need to care about.

So that's why we have these two different but related concepts now.

As mentioned earlier, I added a bunch of JSDoc comments, also here, so hopefully that becomes more clear now!

File: ember-headless-form/src/components/headless-form.ts (hunk 5 above)
Comment (simonihmig) [reply ember-headless-form/src/components/headless-form.ts:298]: Good catch! So I can certainly add `isDestroy{ing,ed}` guards here, even if it's just for code style or best practices.

But I'd first like to understand why this could be a problem here, i.e. what is unsafe about it? 

So when the async action still has a pointer to `this` while the component that `this` refers to is already "destroyed" (unmounted), then `this` will still exist. There is no way it could be "gone", because that's how Garbage Collection works: here our async action is basically the retainer for that object preventing GC.

So it exists, but it could be in a "destroyed" state. In classic Ember, a call to `set(this,...)` or `this.set()` would fail, so yes, that would be bad. But in Octane, would mutating a tracked property be so bad? I don't think anything will throw here, right? And when our form is already "unmounted" (in React's terminology, do we have a better term, unrendered?), this tracked property will just not be consumed anymore, so also not cause rerenders, but that should be all, no further harm caused, or am I missing something?

Compare that to [this comment](https://github.com/ember-cli/eslint-plugin-ember/issues/418#issuecomment-1009181665) here on the linting issue...

Loosely related: should we wrap the async methods (that call potentially async validators) with `waitFor()` from `@ember/test-waiters`, so there are able to cover async things happening that Ember does not auto-detect as async? I would think so, yes?

**Review by ynotdraw — APPROVED** [review]
Code looks great to me! I'm approving, but also curious about the `isDestroy{ing,ed}` comments above so going to track those.

File: ember-headless-form/src/components/-private/types.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nicolechung) [reply ember-headless-form/src/components/-private/types.ts:58]: Thanks for this!

File: test-app/tests/integration/components/headless-form-validation-test.gts (hunk 9 above)
[the remark this answers is not in the document]
Comment (nicolechung) [reply test-app/tests/integration/components/headless-form-validation-test.gts:1138]: Sounds good to me, let's leave it as is

File: ember-headless-form/src/components/headless-form.ts (hunk 5 above)
Comment (NullVoxPopuli) [reply ember-headless-form/src/components/headless-form.ts:298]: >  I would think so, yes?

yeah, that's a good way to solve this whole class of problem.

> what is unsafe about it?

you raise some good points, and I'm going to have to do some concrete research with shareable examples before continuing on with the lint 😅

Comment (simonihmig) [reply ember-headless-form/src/components/headless-form.ts:298]: > accessing services that is the problem in this situation, since service instantiation is lazy

Oh right, that can be problematic indeed!

Comment (simonihmig) [reply ember-headless-form/src/components/headless-form.ts:300]: So I added `@waitFor` to cover async validations, and made the test cover async validations properly (i.e. they were indeed failing without that `@waitFor`), see [this commit](https://github.com/CrowdStrike/ember-headless-form/pull/25/commits/0fb2cec919c1b5a85124c6bb01e2e9e3d89dae5e)

Regarding the `isDestroy{ing,ed}` topic, should we address this here and now, or eventually iterate on it when it's more clear how to address this?

In other words: is this good to merge? @NullVoxPopuli @nicolechung

Comment (NullVoxPopuli) [reply ember-headless-form/src/components/headless-form.ts:300]: we're good to merge. 

isDetroy{ing,ed}, I think I want to iterate on the lint, so we can have a definite error case to match on, and then provide an autofix as well

**Review by NullVoxPopuli — APPROVED** [review]

**Review by nicolechung — APPROVED** [review]
